File: src/builder/Corridors.ts
import {CellBuilder, OPENSPACE, BLOCK_CORR, STAIRS, containsAny, DOORSPACE} from './Cells';
import {CellAttribute, CorridorLayout, TunnelCheck, typedKeys} from '../types/Types';
import { ICorridorBuilder } from '../types/Builders';
import { Direction, DirectionType } from '../types/LayoutTypes';
import { MutableCell } from "../types/MutableTypes";

class CorridorBuilder implements ICorridorBuilder {
    private readonly random: () => number;
    private readonly cellBuilder: CellBuilder;
    private readonly nOddRows: number;
    private readonly nOddCols: number;
    private readonly corridorLayout: CorridorLayout;

    constructor(randomFactory: () => number, cellBuilder: CellBuilder, nOddRows: number, nOddCols: number, corridorLayout: CorridorLayout) {
        this.random = randomFactory;
        this.cellBuilder = cellBuilder;
        this.nOddRows = nOddRows;
        this.nOddCols = nOddCols;
        this.corridorLayout = corridorLayout;
    }

    checkTunnel(r: number, c: number, check: TunnelCheck): boolean {
        return check.corridor.every(([dr, dc]) =>
            this.cellBuilder.cells[r + dr][c + dc].attributes.has(CellAttribute.CORRIDOR)
        ) && check.walled.every(([dr, dc]) =>
            !(containsAny(this.cellBuilder.cells[r + dr]?.[c + dc]?.attributes ?? OPENSPACE, OPENSPACE))
        );
    }

    collapseTunnels(p: number): void {
        if (p === 0) return;
        const removeAll = p === 100;

        for (let i = 0; i < this.nOddRows; i++) {
            const r = i * 2 + 1;
            for (let j = 0; j < this.nOddCols; j++) {
                const c = j * 2 + 1;

                if (containsAny(this.cellBuilder.cells[r][c].attributes, OPENSPACE) && !containsAny(this.cellBuilder.cells[r][c].attributes, STAIRS)) {
                    if (removeAll || this.random() * 100 < p) {
                        this.collapse(r, c);
                    }
                }
            }
        }
    }

    corridors(): void {
        for (let i = 1; i < this.nOddRows; i++) {
            const r = i * 2 + 1;
            for (let j = 1; j < this.nOddCols; j++) {
                const c = j * 2 + 1;

                if (!this.cellBuilder.cells[r][c].attributes.has(CellAttribute.CORRIDOR)) {
                    this.tunnel(i, j);
                }
            }
        }
    }

    private tunnel(i: number, j: number, lastDir: DirectionType | null = null): void {
        const directions = this.tunnelDirs(lastDir);

        for (let dir of directions) {
            if (this.openTunnel(i, j, dir)) {
                const nextI = i + Direction[dir].delta[0];
                const nextJ = j + Direction[dir].delta[1];
                this.tunnel(nextI, nextJ, dir);
            }
        }
    }

    private tunnelDirs(lastDir: DirectionType | null): DirectionType[] {
        const p: number = this.corridorLayout;
        const dirs: DirectionType[] = typedKeys(Direction).sort(() => this.random() - 0.5);

        if (lastDir !== null && this.random() * 100 < p) {
            return [lastDir, ...dirs.filter(d => d !== lastDir)];
        } else {
            return dirs;
        }
    }

    private openTunnel(i: number, j: number, dir: DirectionType): boolean {
        const thisR = i * 2 + 1;
        const thisC = j * 2 + 1;
        const nextR = (i + Direction[dir].delta[0]) * 2 + 1;
        const nextC = (j + Direction[dir].delta[1]) * 2 + 1;
        const midR = (thisR + nextR) / 2;
        const midC = (thisC + nextC) / 2;

        if (this.checkTunnelPlacement(midR, midC, nextR, nextC)) {
            this.delveTunnel(thisR, thisC, nextR, nextC);
            return true;
        } else {
            return false;
        }
    }

    private checkTunnelPlacement(midR: number, midC: number, nextR: number, nextC: number): boolean {
        if (nextR < 0 || nextR > this.cellBuilder.nRows || nextC < 0 || nextC > this.cellBuilder.nCols) return false;

        for (let r = Math.min(midR, nextR); r <= Math.max(midR, nextR); r++) {
            for (let c = Math.min(midC, nextC); c <= Math.max(midC, nextC); c++) {
                if (containsAny(this.cellBuilder.cells[r][c].attributes, BLOCK_CORR)) return false;
            }
        }
        return true;
    }

    private delveTunnel(thisR: number, thisC: number, nextR: number, nextC: number): boolean {
        for (let r = Math.min(thisR, nextR); r <= Math.max(thisR, nextR); r++) {
            for (let c = Math.min(thisC, nextC); c <= Math.max(thisC, nextC); c++) {
                this.cellBuilder.cells[r][c].attributes.delete(CellAttribute.ENTRANCE);
                this.cellBuilder.cells[r][c].attributes.add(CellAttribute.CORRIDOR);
            }
        }
        return true;
    }

    private collapse(r: number, c: number): void {
        if (!containsAny(this.cellBuilder.cells[r][c].attributes, OPENSPACE)) return;

        for (let dir of typedKeys(Direction)) {
            const check = Direction[dir].closeEnd;
            if (this.checkTunnel(r, c, check)) {
                check.close.forEach(([dr, dc]) => {
                    this.cellBuilder.cells[r + dr][c + dc] = new MutableCell();
                });
                const [dr, dc] = check.recurse;
                this.collapse(r + dr, c + dc);
            }
        }
    }
}

export { CorridorBuilder };

File: src/builder/Corridors.kt
package net.bernerbits.dnd.donjon

import java.util.EnumSet
import kotlin.random.Random

class CorridorBuilder(
  private val random: Random,
  private val cellBuilder: CellBuilder,
  private val nI: Int,
  private val nJ: Int,
  private val corridorLayout: CorridorLayout
) : ICorridorBuilder {

  override fun checkTunnel(r: Int, c: Int, check: TunnelCheck): Boolean {
    return check.corridor.all { (dr, dc) -> cellBuilder.cells[r + dr][c + dc].attributes == EnumSet.of(CellAttribute.CORRIDOR) } &&
      check.walled.all { (dr, dc) -> !(cellBuilder.cells.getOrNull(r + dr)?.getOrNull(c + dc)?.attributes?.containsAny(OPENSPACE) ?: true) }
  }

  override fun collapseTunnels(p: Int) {
    if (p == 0) return
    val removeAll = p == 100

    for (i in 0 until nI) {
      val r = (i * 2) + 1
      for (j in 0 until nJ) {
        val c = (j * 2) + 1

        if (cellBuilder.cells[r][c].attributes.containsAny(OPENSPACE) && !cellBuilder.cells[r][c].attributes.containsAny(STAIRS)) {
          if (removeAll || random.nextInt(100) < p) {
            collapse(r, c)
          }
        }
      }
    }
  }

  override fun corridors() {
    for (i in 1 until nI) {
      val r = (i * 2) + 1
      for (j in 1 until nJ) {
        val c = (j * 2) + 1

        if (!cellBuilder.cells[r][c].attributes.contains(CellAttribute.CORRIDOR)) {
          tunnel(i, j)
        }
      }
    }
  }

  private fun tunnel(i: Int, j: Int, lastDir: Direction? = null) {
    val directions = tunnelDirs(lastDir)

    for (dir in directions) {
      if (openTunnel(i, j, dir)) {
        val nextI = i + dir.delta.first
        val nextJ = j + dir.delta.second
        tunnel(nextI, nextJ, dir)
      }
    }
  }

  private fun tunnelDirs(lastDir: Direction?): List<Direction> {
    val p = corridorLayout.probability
    val dirs = Direction.entries.shuffled()

    return if (lastDir != null && random.nextInt(100) < p) {
      listOf(lastDir) + dirs.filter { it != lastDir }
    } else {
      dirs
    }
  }

  private fun openTunnel(i: Int, j: Int, dir: Direction): Boolean {
    val thisR = (i * 2) + 1
    val thisC = (j * 2) + 1
    val nextR = ((i + dir.delta.first) * 2) + 1
    val nextC = ((j + dir.delta.second) * 2) + 1
    val midR = (thisR + nextR) / 2
    val midC = (thisC + nextC) / 2

    return if (checkTunnelPlacement(midR, midC, nextR, nextC)) {
      delveTunnel(thisR, thisC, nextR, nextC)
    } else false
  }

  private fun checkTunnelPlacement(midR: Int, midC: Int, nextR: Int, nextC: Int): Boolean {
    if (nextR < 0 || nextR > cellBuilder.nRows || nextC < 0 || nextC > cellBuilder.nCols) return false

    val rRange = minOf(midR, nextR)..maxOf(midR, nextR)
    val cRange = minOf(midC, nextC)..maxOf(midC, nextC)

    for (r in rRange) {
      for (c in cRange) {
        if (cellBuilder.cells[r][c].attributes.any { it in BLOCK_CORR }) return false
      }
    }
    return true
  }

  private fun delveTunnel(thisR: Int, thisC: Int, nextR: Int, nextC: Int): Boolean {
    val rRange = minOf(thisR, nextR)..maxOf(thisR, nextR)
    val cRange = minOf(thisC, nextC)..maxOf(thisC, nextC)

    for (r in rRange) {
      for (c in cRange) {
        cellBuilder.cells[r][c].attributes.remove(CellAttribute.ENTRANCE)
        cellBuilder.cells[r][c].attributes.add(CellAttribute.CORRIDOR)
      }
    }
    return true
  }

  private fun collapse(r: Int, c: Int) {
    if (!cellBuilder.cells[r][c].attributes.containsAny(OPENSPACE)) return

    for (dir in Direction.entries) {
      val check = dir.closeEnd
      if (checkTunnel(r, c, check)) {
        check.close.forEach { (dr, dc) ->
          cellBuilder.cells[r + dr][c + dc] = MutableCell()
        }
        check.recurse.let { (dr, dc) -> collapse(r + dr, c + dc) }
      }
    }
  }

}
